src/uct/ib/mlx5/dv/ib_mlx5dv_md.c

```diff
@@ -858,6 +858,313 @@ uct_ib_mlx5_devx_dereg_keys(uct_ib_mlx5_md_t *md, uct_ib_mlx5_devx_mem_t *memh)
+    qp_attr.max_dest_rd_atomic = 1;
+    qp_attr.ah_attr.port_num   = port_num;
+    qp_attr.ah_attr.dlid       = port_attr->lid;
+    qp_attr.ah_attr.is_global  = 1;
+
+    if (UCS_OK != uct_ib_device_query_gid(ibdev, port_num,
+                                          UCT_IB_DEVICE_DEFAULT_GID_INDEX,
+                                          &qp_attr.ah_attr.grh.dgid,
+                                          UCS_LOG_LEVEL_ERROR)) {
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ret = ibv_modify_qp(md->umr_qp, &qp_attr, attr_mask);
+    if (ret) {
+        ucs_error("%s: failed to modify UMR QP (0x%x) to RTR: %m",
+                  uct_ib_device_name(ibdev), md->umr_qp->qp_num);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    /* Modify to RTS */
+    attr_mask             = IBV_QP_STATE |
+                            IBV_QP_SQ_PSN |
+                            IBV_QP_TIMEOUT |
+                            IBV_QP_RNR_RETRY |
+                            IBV_QP_RETRY_CNT |
+                            IBV_QP_MAX_QP_RD_ATOMIC;
+    qp_attr.qp_state      = IBV_QPS_RTS;
+    qp_attr.sq_psn        = 0;
+    qp_attr.timeout       = 7;
+    qp_attr.rnr_retry     = 7;
+    qp_attr.retry_cnt     = 7;
+    qp_attr.max_rd_atomic = 1;
+
+    ret = ibv_modify_qp(md->umr_qp, &qp_attr, attr_mask);
+    if (ret) {
+        ucs_error("%s: failed to modify UMR QP (0x%x) to RTS: %m",
+                  uct_ib_device_name(ibdev), md->umr_qp->qp_num);
+        return UCS_ERR_IO_ERROR;
+    }
+
+    ucs_trace("%s: initialized UMR QP 0x%x", uct_ib_device_name(ibdev),
+              md->umr_qp->qp_num);
+
+    return UCS_OK;
+}
+
+static ucs_status_t uct_ib_mlx5_devx_umr_init(uct_ib_mlx5_md_t *md)
+{
+    ucs_status_t status;
+
+    if ((status = uct_ib_mlx5_devx_umr_create_cq(md)) != UCS_OK) {
+        goto err;
+    }
+
+    if ((status = uct_ib_mlx5_devx_umr_create_qp(md)) != UCS_OK) {
+        goto err_destroy_cq;
+    }
+
+    if ((status = uct_ib_mlx5_devx_umr_modify_qp(md)) != UCS_OK) {
+        goto err_destroy_qp;
+    }
+
+    /* Initialize indirect UMR mkey pool */
+    ucs_list_head_init(&md->umr_mkey_pool);
+
+    /* Initialize indirect UMR mkey hash table */
+    md->umr_mkey_hash = kh_init(umr_mkey_map);
+    if (md->umr_mkey_hash == NULL) {
+        status = UCS_ERR_NO_MEMORY;
+        goto err_destroy_qp;
+    }
+
+    return UCS_OK;
+
+err_destroy_qp:
+    uct_ib_destroy_qp(md->umr_qp);
+err_destroy_cq:
+    ibv_destroy_cq(md->umr_cq);
+err:
+    return status;
+}
+
+static void uct_ib_mlx5_devx_umr_free(uct_ib_mlx5_md_t *md)
+{
+    uct_ib_device_t *ibdev = &md->super.dev;
+    uct_ib_mlx5_devx_umr_alias_t mkey_alias;
+    uct_ib_mlx5_devx_umr_t *item, *tmp;
+    ucs_status_t status;
+
+    if (NULL != md->umr_mkey_hash) {
+        if (kh_size(md->umr_mkey_hash) != 0) {
+            ucs_info("%s: UMR mkey hash has %d elements",
+                     uct_ib_device_name(ibdev), kh_size(md->umr_mkey_hash));
+        }
```

Comment:
it will always be non-empty, no need to print it in info level (anyway you will print each entry in trace)

Reply:
No, it's not necessarily non-empty.
UMR hash map is populated only when process imports UMR key (e.g. DPU offloading daemon side).

But this hash map object is unconditionally created on all the MD devices supporting indirect xgvmi, even on the exporter side, even if not really using xgvmi feature at all. In all those cases, UMR hash map remains empty.

Why need this print - well, just for debugging purposes. When daemon is terminated it's useful to see the amount of mkeys imported.
Maybe log level can be lowered to debug or trace?

Reply:
for now I lowered log level to trace

Reply:
Well, it was useful for me during development phase - so that I can quickly compare the amount of keys created on exporter and received on importer. If we have only log entry per element - it would be harder to compare.

If you don't like extra `if`, maybe we just print trace unconditionally:
```
    if (NULL != md->umr_mkey_hash) {
        ucs_trace("%s: UMR mkey hash has %d elements",
                  uct_ib_device_name(ibdev), kh_size(md->umr_mkey_hash));
```
??

Reply:
ok